`multiline_arguments_brackets` requires last tailing closure to contain a newline

### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

It seems like the `multiline_arguments_brackets` assumes that the end of the `callBody` is still within the parenthesis for the arguments.

In the case where call contains a trailing closure, the rule will generate a false positive when the closure is a single line.

In the case where the call contains multiple trailing closure, the rule will generate a false positive when the last trailing closure is a single line. Earlier trailing closures can be one one or multiple lines and it wont affect the results.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
➜ swiftlint lint example.swift --no-cache --enable-all-rules 
Linting Swift files at paths example.swift
Linting 'example.swift' (1/1)
The `anyobject_protocol` rule is now deprecated and will be completely removed in a future release.
/Users/neo/Desktop/example.swift:3:25: warning: Multiline Arguments Brackets Violation: Multiline arguments should have their surrounding brackets in a new line. (multiline_arguments_brackets)
Done linting! Found 1 violation, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.50.0
* Installation method used (Homebrew, CocoaPods, building from source, etc)? home-brew
* Paste your configuration file: n/a
* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? 14.1
* Do you have a sample that shows the issue? Yes

```zsh
echo "SomeType(\n    a: 1\n) { print("completion") }" | swiftlint lint --no-cache --use-stdin --enable-all-rules
```

```swift
SomeType(
    a: 1
) { print("completion") }
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -322,6 +322,11 @@
   [SimplyDanny](https://github.com/SimplyDanny)
   [#4234](https://github.com/realm/SwiftLint/issues/4234)
 
+* Fix false-positives from `multiline_arguments_brackets` when a function call has a
+  single line trailing closure.  
+  [CraigSiemens](https://github.com/CraigSiemens)
+  [#4510](https://github.com/realm/SwiftLint/issues/4510)
+
 ## 0.49.1: Buanderie Principale
 
 _Note: The default branch for the SwiftLint git repository was renamed from

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -128,6 +128,7 @@
   - `legacy_objc_type`
   - `legacy_random`
   - `lower_acl_than_parent`
+  - `multiline_arguments_brackets`
   - `multiline_parameters`
   - `multiple_closures_with_trailing_closure`
   - `no_extension_access_modifier`

--- a/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
@@ -1,7 +1,7 @@
 import Foundation
-import SourceKittenFramework
+import SwiftSyntax
 
-public struct MultilineArgumentsBracketsRule: ASTRule, OptInRule, ConfigurationProviderRule {
+public struct MultilineArgumentsBracketsRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule {
     public var configuration = SeverityConfiguration(.warning)
 
     public init() {}
@@ -159,68 +159,45 @@ public struct MultilineArgumentsBracketsRule: ASTRule, OptInRule, ConfigurationP
         ]
     )
 
-    public func validate(file: SwiftLintFile,
-                         kind: SwiftExpressionKind,
-                         dictionary: SourceKittenDictionary) -> [StyleViolation] {
-        guard
-            kind == .call,
-            let bodyRange = dictionary.bodyByteRange,
-            let range = file.stringView.byteRangeToNSRange(bodyRange)
-        else {
-            return []
-        }
+    public func makeVisitor(file: SwiftLintFile) -> ViolationsSyntaxVisitor {
+        Visitor(viewMode: .sourceAccurate)
+    }
+}
 
-        let callBody = file.contents.substring(from: range.location, length: range.length)
+private extension MultilineArgumentsBracketsRule {
+    final class Visitor: ViolationsSyntaxVisitor {
+        override func visitPost(_ node: FunctionCallExprSyntax) {
+            guard let firstArgument = node.argumentList.first,
+                  let leftParen = node.leftParen,
+                  let rightParen = node.rightParen else {
+                return
+            }
 
-        let parameters = dictionary.substructure.filter {
-            // Argument expression types that can contain newlines
-            [.argument, .array, .dictionary, .closure, .call].contains($0.expressionKind)
-        }
-        let parameterBodies = parameters.compactMap { $0.content(in: file) }
-        let parametersNewlineCount = parameterBodies.map { body in
-            return body.countOccurrences(of: "\n")
-        }.reduce(0, +)
-        let callNewlineCount = callBody.countOccurrences(of: "\n")
-        let isMultiline = callNewlineCount > parametersNewlineCount
+            let hasMultilineFirstArgument = hasLeadingNewline(firstArgument)
+            let hasMultilineArgument = node.argumentList
+                .contains { argument in
+                    hasLeadingNewline(argument)
+                }
 
-        guard isMultiline else {
-            return []
-        }
+            let hasMultilineRightParen = hasLeadingNewline(rightParen)
 
-        let trailingClosurePattern = "\\) \\{[^\\}]*\\z"
-        let trailingClosureRegex = regex(trailingClosurePattern)
+            if !hasMultilineFirstArgument, hasMultilineArgument {
+                violations.append(leftParen.endPosition)
+            }
 
-        let expectedBodyBeginRegex = regex("\\A(?:[ \\t]*\\n|[^\\n]*(?:in|\\{)\\n)")
-        let expectedBodyEndRegex = regex("\\n[ \\t]*\\z")
-        let expectedTrailingClosureBodyEndRegex = regex("\\n[ \\t]*\\) \\{.*\\z")
+            if !hasMultilineArgument, hasMultilineRightParen {
+                violations.append(leftParen.endPosition)
+            }
 
-        // Should only ever be ")" when the call doesn't have a trailing closure, or "}" when it does.
-        let followingCharacter = file.contents.substring(from: range.location + range.length, length: 1)
-
-        var violatingByteOffsets = [ByteCount]()
-        if expectedBodyBeginRegex.firstMatch(in: callBody, options: [], range: callBody.fullNSRange) == nil {
-            violatingByteOffsets.append(bodyRange.location)
-        }
-
-        if followingCharacter == ")",
-           expectedBodyEndRegex.firstMatch(in: callBody, range: callBody.fullNSRange) == nil {
-            violatingByteOffsets.append(bodyRange.upperBound)
-        } else if followingCharacter == "}",
-                  expectedTrailingClosureBodyEndRegex.firstMatch(in: callBody, range: callBody.fullNSRange) == nil {
-            if let match = trailingClosureRegex.firstMatch(in: callBody, range: callBody.fullNSRange) {
-                let matchFileLocation = range.location + match.range.location
-                let offset = file.stringView.byteOffset(fromLocation: matchFileLocation)
-                violatingByteOffsets.append(offset)
-            } else {
-                violatingByteOffsets.append(bodyRange.upperBound)
+            if !hasMultilineRightParen, hasMultilineArgument {
+                violations.append(rightParen.position)
             }
         }
 
-        return violatingByteOffsets.map { byteOffset in
-            StyleViolation(
-                ruleDescription: Self.description, severity: configuration.severity,
-                location: Location(file: file, byteOffset: byteOffset)
-            )
+        private func hasLeadingNewline(_ syntax: SyntaxProtocol) -> Bool {
+            guard let leadingTrivia = syntax.leadingTrivia else { return false }
+
+            return leadingTrivia.pieces.contains { $0.isNewline }
         }
     }
 }

--- a/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
@@ -72,6 +72,35 @@ public struct MultilineArgumentsBracketsRule: ASTRule, OptInRule, ConfigurationP
             SomeType(a: [
                 1, 2, 3
             ], b: [1, 2])
+            """),
+            Example("""
+            SomeType(
+              a: 1
+            ) { print("completion") }
+            """),
+            Example("""
+            SomeType(
+              a: 1
+            ) {
+              print("completion")
+            }
+            """),
+            Example("""
+            SomeType(
+              a: .init() { print("completion") }
+            )
+            """),
+            Example("""
+            SomeType(
+              a: .init() {
+                print("completion")
+              }
+            )
+            """),
+            Example("""
+            SomeType(
+              a: 1
+            ) {} onError: {}
             """)
         ],
         triggeringExamples: [
@@ -112,6 +141,16 @@ public struct MultilineArgumentsBracketsRule: ASTRule, OptInRule, ConfigurationP
                 b: "two"↓)
             """),
             Example("""
+            SomeOtherType(
+              a: 1↓) {}
+            """),
+            Example("""
+            SomeOtherType(
+              a: 1↓) {
+              print("completion")
+            }
+            """),
+            Example("""
             views.append(ViewModel(
                 title: "MacBook", subtitle: "M1", action: { [weak self] in
                 print("action tapped")
@@ -148,16 +187,33 @@ public struct MultilineArgumentsBracketsRule: ASTRule, OptInRule, ConfigurationP
             return []
         }
 
+        let trailingClosurePattern = "\\) \\{[^\\}]*\\z"
+        let trailingClosureRegex = regex(trailingClosurePattern)
+
         let expectedBodyBeginRegex = regex("\\A(?:[ \\t]*\\n|[^\\n]*(?:in|\\{)\\n)")
         let expectedBodyEndRegex = regex("\\n[ \\t]*\\z")
+        let expectedTrailingClosureBodyEndRegex = regex("\\n[ \\t]*\\) \\{.*\\z")
+
+        // Should only ever be ")" when the call doesn't have a trailing closure, or "}" when it does.
+        let followingCharacter = file.contents.substring(from: range.location + range.length, length: 1)
 
         var violatingByteOffsets = [ByteCount]()
         if expectedBodyBeginRegex.firstMatch(in: callBody, options: [], range: callBody.fullNSRange) == nil {
             violatingByteOffsets.append(bodyRange.location)
         }
 
-        if expectedBodyEndRegex.firstMatch(in: callBody, options: [], range: callBody.fullNSRange) == nil {
+        if followingCharacter == ")",
+           expectedBodyEndRegex.firstMatch(in: callBody, range: callBody.fullNSRange) == nil {
             violatingByteOffsets.append(bodyRange.upperBound)
+        } else if followingCharacter == "}",
+                  expectedTrailingClosureBodyEndRegex.firstMatch(in: callBody, range: callBody.fullNSRange) == nil {
+            if let match = trailingClosureRegex.firstMatch(in: callBody, range: callBody.fullNSRange) {
+                let matchFileLocation = range.location + match.range.location
+                let offset = file.stringView.byteOffset(fromLocation: matchFileLocation)
+                violatingByteOffsets.append(offset)
+            } else {
+                violatingByteOffsets.append(bodyRange.upperBound)
+            }
         }
 
         return violatingByteOffsets.map { byteOffset in

--- a/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/MultilineArgumentsBracketsRule.swift
@@ -1,4 +1,3 @@
-import Foundation
 import SwiftSyntax
 
 public struct MultilineArgumentsBracketsRule: SwiftSyntaxRule, OptInRule, ConfigurationProviderRule {

